sources/api/early-boot-config/src/provider/vmware.rs

```diff
@@ -125,23 +134,146 @@ impl VmwareDataProvider {
 
         Ok(decoded)
     }
+
+    /// Read and decode user data based on values retrieved from the guestinfo interface
+    fn guestinfo_user_data() -> Result<Option<SettingsJson>> {
+        info!("Attempting to retrieve user data via guestinfo interface");
+
+        // `guestinfo.userdata.encoding` informs us how to handle the data in the
+        // `guestinfo.userdata` field
+        let maybe_encoding = Self::backdoor_get_bytes(Self::GUESTINFO_USERDATA_ENCODING)?;
+        let user_data_encoding: UserDataEncoding = match maybe_encoding {
+            Some(val) => {
+                let encoding_str = String::from_utf8(val).context(error::InvalidUtf8 {
+                    what: Self::GUESTINFO_USERDATA_ENCODING,
+                })?;
+                info!("Found user data encoding: {}", encoding_str);
+
+                serde_plain::from_str(&encoding_str).context(error::UnknownEncoding {
+                    encoding: encoding_str,
+                })?
+            }
+
+            // The cloudinit VMware guestinfo data provider assumes any user data without an
+            // associated encoding means raw data is being passed.  We will follow suit here.
+            None => {
+                warn!(
+                    "'{}' unset, assuming raw user data",
+                    Self::GUESTINFO_USERDATA_ENCODING
+                );
+                UserDataEncoding::Raw
+            }
+        };
+
+        let user_data_bytes = match Self::backdoor_get_bytes(Self::GUESTINFO_USERDATA)? {
+            Some(val) => val,
+            None => return Ok(None),
+        };
+
+        let user_data_string = match user_data_encoding {
+            // gzip+base64 is gzip'ed user data that is base64 encoded
+            UserDataEncoding::Base64 | UserDataEncoding::GzipBase64 => {
+                info!("Decoding user data");
+                let mut reader = Cursor::new(user_data_bytes);
+                let mut decoder = base64::read::DecoderReader::new(&mut reader, base64::STANDARD);
+
+                let mut decoded_bytes = Vec::new();
+                decoder
+                    .read_to_end(&mut decoded_bytes)
```

Comment:
It seems like it'd be simpler and more efficient to wrap an OptionalCompressionReader around a DecoderReader, rather than decoding base64 into a string and then expanding that?